design/aaa.md

```diff
@@ -0,0 +1,175 @@
+...
+DEBUG  [2021-09-27T09:46:50+03:00]pkg/api/controller.go:2912 pkg/api.(*Controller).LogAction performing API action                         action=list_policies host="localhost:8000" message_type=action method=GET path="/api/v1/auth/policies?prefix=&after=&amount=100" request_id=ae2fbb59-f54f-4cb4-a25c-7dd1d7280538 service=api_gateway service_name=rest_api
+```
+
+A usable audit log would require at least:
+
+1. Identities not credential IDs.
+1. Identities linked to actions.
+1. Explicit authentication events linking identities to credentials used
+   to authenticate.
+
+## Active Directory and LDAP support
+
+Issue #2058 is to support Active Directory.  This splits naturally along
+the 3 A's: use Active Directory to _authenticate_, to _authorize_, or to
+receive _audit_ logs of user operations.
+
+### Authentication
+
+This section will initially apply to GUI clients.  We will have to allow
+users to request access keys (or short-term tokens) to allow them to run
+programs that use AWS S3 or the lakeFS API.
+
+We may authenticate using an LDAP server directly or using some external
+application.
+
+* To authenticate directly, the lakeFS web server should use SASL.  That
+  may also allow other methods in future.
+  
+  The advantage here is that *all* LDAP capabilities can be used as long
+  as lakeFS understands them.  This probably helps most in enterprises.
+
+* To use external applications, the lakeFS web server should use OAuth2.
+
+  The advantage here is that other OAuth2 providers can be used: GitHub,
+  Google Suites, and many others.  LDAP users can still use bridges such
+  as Dex to authenticate over OAuth2.  The [Microsoft Connector][dex-ms]
+  gives an example.
+
+#### Decision
+
+Add OAuth2 support, as it provides easier integration with more identity
+providers by using bridges.  Allow users who authenticate in this way to
+create access keys for their accounts, allowing them to use applications
+over both the S3 gateway and the lakeFS API.
+
+The lakeFS web UI will (optionally) authenticate users using OAuth2, and
+place the result on their JWT.  Selecting between OAuth2 or internal can
+be done once during the initial login.  After that the web app can reuse
+that method.
+
+### Authorization
+
+There is room for business login when authorizing.  Typically users have
+properties attached during authentication, which some business logic can
+connect to attached policies.  This can be quite complex, but we already
+have a groups mechanism in place which we might re-use.
+
+When do we attach policies do these groups?  Configuration changes apply
+at different times depending on this decision.  Authenticated users will
+see only changes in mappings that are attached before the change.
+
+#### Decision
+
+lakeFS shall receive groups from OAuth2 and encode them as claims on the
+JWT.  The remaining logic can proceed as today.
```

Comment:
Would the groups be added to our IAM groups automatically/manually by the admin? I'm missing the full picture of setting up the oauth app, configuring the groups, attaching to external groups, etc..